### PR TITLE
Kapa refactor

### PR DESCRIPTION
**Pull request by tcezard**
This PR creates a new framework for validation of StepEPPs
It also adds a new object for automatic generation of fake Entities to use in tests.

### CODE REVIEW

File: EPPs/common.py [hunk 1]
```diff
@@ -144,6 +153,86 @@ def run(self):
...
+    @cached_property
+    def input_container_names(self):
+        """The name of containers from input artifacts. """
+        containers = set()
+        for art in self.artifacts:
+            # Check to see if artifact has a container before retrieving the container.
+            # Artifacts that are not samples will not have containers.
+            if art.container:
+                containers.add(art.container.name)
+        return list(containers)
```
Comment (mwhamgenomics) [inline EPPs/common.py]: Might we end up with non-deterministic order here?

File: EPPs/common.py [hunk 2]
```diff
@@ -144,6 +153,86 @@ def run(self):
...
+         - _max_nb_input: set the maximum number of input of the step.
+         - _nb_analyte_per_input: set the number of replicates (Analytes) required for the step.
+         - _nb_resfile_per_input: set the number of replicates (ResultFiles) required for the step.
+         - _max_nb_project: set the maximum number of project in the step.
+        """
+        if self._max_nb_input_containers is not None:
+            # check the number of input containers
+            if len(self.input_container_names) > self._max_nb_input_containers:
+                raise InvalidStepError(
+                    message='Maximum number of input container is %s. There are %s input container in the step.' % (
```
Comment (mwhamgenomics) [inline EPPs/common.py]: [This requested change](https://github.com/EdinburghGenomics/clarity_scripts/pull/69#discussion_r232696073) is still not fully implemented.

Comment (tcezard) [reply EPPs/common.py]: `InvalidStepError` now inherit from `Exception` and the parameter `message` is required to make it mandatory and explicit to provide a message to be display.

File: EPPs/common.py (hunk 1 above)
Comment (tcezard) [reply EPPs/common.py]: Should be deterministic now

File: EPPs/common.py [hunk 3]
```diff
@@ -31,6 +31,14 @@ class StepEPP(app_logging.AppLogger):
     _process = None
     _use_load_config = True
 
+    # Step Validation parameters
+    _max_nb_input_containers = None
+    _max_nb_output_containers = None
+    _max_nb_input = None
+    _nb_analyte_per_input = None
+    _nb_resfile_per_input = None
+    _max_nb_project = None
```
[the remark this answers is not in the document]
Comment (tcezard) [reply EPPs/common.py]: should be plural

File: scripts/autoplacement_seq_plate_quant.py [hunk 4]
```diff
@@ -25,23 +26,19 @@ def _run(self):
...
-            outputs = sorted(self.process.outputs_per_input(art.id,ResultFile=True), key=lambda x: x.name, reverse=False)
+        for art in self.artifacts:
 
-            if len(outputs) != 3:
-                raise InvalidStepError(
-                    message="3 replicates required for each sample and standard. Did you remember to click 'Apply' when assigning replicates?")
+            # obtain list outputs for the inputs. Sort list in case each output replicate has a unique name based on
+            # the replicate number e.g. sampleAreplicate-1, sampleAreplicate-2
+            outputs = sorted(self.process.outputs_per_input(art.id, ResultFile=True), key=lambda x: x.name,
+                             reverse=False)
```
[the remark this answers is not in the document]
Comment (tcezard) [reply scripts/autoplacement_seq_plate_quant.py]: removed `reverse=False`

File: EPPs/common.py (hunk 2 above)
Comment (mwhamgenomics) [reply EPPs/common.py]: As discussed, the single argument to the constructor makes sure only one argument is passed. It doesn't call the superconstructor, though - `super().__init__` call to be added.

**Pull request merged**